### ISSUE TITLE
[HOLD on #3381] Selecting, copying, and pasting text on web and desktop are weird and in inverted order

### ISSUE THREAD

**Issue by puneetlath** [Expensify/Expensify.cash#1341]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/master/CONTRIBUTING.md) for onboarding!
___
Upwork job:  https://www.upwork.com/jobs/~0195a298f26a002c08
 
**NOTE:** We are **only** looking for proposals on fixing this directly in https://github.com/facebook/react-native.  Example issue addressing this accessibility issue https://github.com/facebook/react-native/issues/30373
We are **not** looking for a work-around in our repo

**Platform - version:**  
Web and Desktop. 

**Action Performed (reproducible steps):**
Attempt to select text across multiple chat inputs. You will find that it is difficult to select text across multiple chat inputs. Then, once you have managed to select multiple lines of text, try pasting that text. You will find that the order of the text has been inverted. 

**Expected Result:**
That it would be easy to select text across chat inputs and that when copying and pasting the text it would retain itss order.

**Actual Result:**
The text is difficult to select and pastes in inverted order. 

**Notes/Photos/Videos:** 
![image](https://user-images.githubusercontent.com/1371996/105929434-2bc36880-600d-11eb-95bc-71393d474f08.png)

 
![image](https://user-images.githubusercontent.com/1371996/105929489-531a3580-600d-11eb-883e-d52d8d05ccb4.png)

We use [Flatlist](https://reactnative.dev/docs/flatlist) for rendering the list of chat messages. We use the `inverted` prop which reverses the scroll direction on mobile. We then use react-native-web for web and desktop. The ideal solution would fix this upstream, either in React Native or react-native-web. You can see an example of the problem here: https://codesandbox.io/s/cold-wave-grzpj?file=/src/App.js

**Comment by parasharrajat** [comment]
So I did some digging. The issue is that inverted Flatlist uses `transform:scale(-1)` for inversion but the Browser selection works based on the order of dom elements. Even if we change the Flatlist code to use `flex-direction:column-reverse` the problem will remain same as the order of dom element is still reversed.

**Comment by puneetlath** [comment]
@parasharrajat can you think of a way FlatList could implement this functionality that wouldn't require inverting the order of the dom elements?

**Comment by puneetlath** [comment]
@anthony-hull pointed out that this also creates an accessibility issue: https://github.com/facebook/react-native/issues/30373

**Comment by parasharrajat** [comment]
@puneetlath This issue is not with the FlatList. It is just a wrapper but more of the VirtualizedList and ScrollView underneath. And this is not really an accessibility issue. FlatList is made for the purpose of mobile devices so inversion logic is correct. Why would they create two-dimensional renderings if they do not need to? So they are not fixing this issue. It's been there for years now.


 Yeah. I was suggesting using a non-inverted FlatList. I have a plan but first I need to give it a go. I think it is possible to show the chats with the non-inverted list. I have found some supporting evidence for my plan. I will soon let you in few hours.

**Comment by marcaaron** [comment]
Gonna add this context here since @joaniew brought to our attention another bug related to this. I have a strong suspicion it is related to the `FlatList` CSS issues (if not the same exact issue) so whenever we fix this we should make sure to check whether this is resolved as well.

**When copying text from Expensify.cash and pasting into gmail we see this:**

<img width="349" alt="screen_shot_2021-02-24_at_4 42 30_pm" src="https://user-images.githubusercontent.com/32969087/109188472-138f5800-7737-11eb-8a3f-1307bf972e94.png">

**Comment by mallenexpensify** [comment]
cc @tgolen so you can follow along

**Comment by henrymoulton** [comment]
Howdy, I'm a React Native contractor/freelancer currently at Amazon.

I think this format of open source core + contract work to solve issues is super cool.

I looked through the issues and source code of the `react-native-web` `FlatList`/`VirtualizedList`.

My current conclusion is that the approach to inversion inside React Native is pretty broken, affecting scroll + text highlight + A11y.

Like @parasharrajat says:
> I was suggesting using a non-inverted FlatList.

I have a screenshot showing that below, but appreciate the outcome you probably want is a React Native fork which would change the `inverted` implementation?

<img width="251" alt="image" src="https://user-images.githubusercontent.com/30093692/112948259-c1739500-912f-11eb-9951-6eaf24dbd8a1.png">

**Comment by mallenexpensify** [comment]
Thanks for taking a look @henrymoulton 
> the outcome you probably want is a React Native fork which would change the inverted implementation?

Yes, we're hoping for a React Native fork so that we can fix the issue for ourselves as well as everyone else! 
Do you think this is something you'd be able to invest the time in to fix?  
If not, we have many other smaller jobs on Upwork - [link.](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

**Comment by iwiznia** [comment]
To clarify, ideally the fix would be merged into the react native code base. So it would be good to also check with the maintainers that the proposed solution is good for them (once we review it here).

**Comment by ahmedjamshed001** [comment]
This issue can be resolved by using modified FlatList component here

https://codesandbox.io/s/musing-rubin-tti3o

**Comment by mallenexpensify** [comment]
Hi @ahmedjamshed001 We're not looking to override FlatList functionality, we want flatlist itself changed. Can you propose a solution for that?

**Comment by nbgspl-github** [comment]
Hi 

  I looked through the issues and source code of the react-native-web FlatList/VirtualizedList.

My current conclusion is that   inversion inside React Native is pretty broken, affecting scroll + text highlight + A11y.

 
As few people suggested above I agreed suggesting using a non-inverted FlatList.

 Thanks
 Sumit NBGSPL

**Comment by mallenexpensify** [comment]
@puneetlath Nirav in Upwork proposed 
>I have reverse order of data instead of element
I had to do change on the other three screens as I have changed the data order in the element
that change is
removed inverted from component and changed y position


https://user-images.githubusercontent.com/22508304/114251545-4479c400-9956-11eb-9726-50af816461fb.mov

Can you provide some feedback to share with Nirav?

**Comment by puneetlath** [comment]
Hm, their solution doesn't look like it solves the underlying FlatList issue, which is the approach we'd like. It looks like they are just suggesting a workaround.

**Comment by marcaaron** [comment]
Based on the discussions in Slack, how do we feel about explicitly removing this requirement:

>The ideal solution would fix this upstream, either in React Native or react-native-web

Yes, the ideal solution should _theoretically_ be something we _could_ merge into the main `react-native` codebase (not `react-native-web` as the problem is not with this package). However, we should be clear that that the solution doesn't _need_ to be merged into `react-native` in order to be considered "done". It just needs to be:

- A modified version of `FlatList` (in it's current state) where the `inverted` prop works correctly and continues to work when it's not `inverted`
- Ideally this version would be offered as a self-contained module so that we can use it alongside `react-native` and not maintain a fork of the entire `react-native` package

Thoughts?

**Comment by tgolen** [comment]
I think that means we would prefer maintaining our own fork as opposed to
getting it fixed in RN, and I don't understand why that would be ideal. It
seems to me that will just end up with a big maintenance cost at some point.

**Comment by tgolen** [comment]
Ah, sorry... I read your comment wrong. I see that you said that we don't want to maintain any forks 👍

**Comment by tgolen** [comment]
I guess that sounds like an alternative that we might reach at some point, but I feel like we haven't quite exhausted all the effort to get this into RN either... @puneetlath Maybe you should take your issue (https://github.com/facebook/react-native/issues/30383) and start contacting their twitter feed or their discord channel about it? It would at least be good to hear from a maintainer what the feasibility of getting this fixed is.

**Comment by iwiznia** [comment]
Yeah, let's try that. We should also clarify that we are willing to sponsor with money and/or engineering time the implementation of the solution.

**Comment by tgolen** [comment]
Yeah, perhaps we can work on this messaging a little bit too so that it looks a little more legit (cc @mallenexpensify ):

![image](https://user-images.githubusercontent.com/1228807/114946871-3f5ebe00-9e09-11eb-8ca5-664af53559d6.png)

Something like:

> Hey @teod, thanks for the comments above. That makes sense about the scroll direction. I'd really like to get this fixed and I was curious if you'd be interested in working on this for me? Of course, I would reimburse you for your time and effort. If you're interested, I've posted a job to UpWork that you can grab.

**Comment by mallenexpensify** [comment]
Yeah, my message is pretty blunt.  I can def drop in another post of someone else can. I'm going to let this marinate til tomorrow before doing anything

**Comment by mallenexpensify** [comment]
Still need to work on this, added `Weekly` so I don't forget.  Comment linking and this GH made me think of it - https://github.com/Expensify/Expensify/issues/147480#issuecomment-827044976 made me

**Comment by mallenexpensify** [comment]
Apologies for the delays, I'm back working 100% this week so I can start to prioritize.. again. 

Where I plan to start
From Tim G
> I feel like we haven't quite exhausted all the effort to get this into RN either... @puneetlath Maybe you should take your issue (facebook/react-native#30383) and start contacting their twitter feed or their discord channel about it? It would at least be good to hear from a maintainer what the feasibility of getting this fixed is.

From Marc
> Try to reach out directly to react native maintainers to see if they respond or have any ideas about whether it's realistic to fix upstream ?

> People who have worked on VirtualizedList in the past 6 months with public emails or socials
https://github.com/jimmy623
https://github.com/Simek
https://github.com/kacieb
https://github.com/yungsters
https://github.com/sasurau4
also https://www.facebook.com/spencer wrote a lot of the code

> I'd maybe approach those conversations from the perspective of seeing if they will offer free insights into whether what we want to do is possible.

**Comment by mallenexpensify** [comment]
Created a new, clean issue for this with the deliverable being
> The deliverable is to fix the core issue. The PR should be submitted against our RN fork (Expensify/react-native).

View it here - https://github.com/Expensify/Expensify.cash/issues/3381
I'm putting this issue on hold. Once 3381 is finished the plan is to 
- Implement it to fix our issue
- Try to get RN to merge the fork

**Comment by mallenexpensify** [comment]
Closing this, once we're able to fix https://github.com/Expensify/Expensify.cash/issues/3381 we can create a new issue with/for next steps
